presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1348,11 +1356,27 @@ else if (expression instanceof LongLiteral) {
                     }
 
                     // otherwise, just use the expression as is
+                    boolean orderByOnExpression = false;
                     if (orderByExpression == null) {
                         orderByExpression = expression;
+                        orderByOnExpression = true;
                     }
 
                     ExpressionAnalysis expressionAnalysis = analyzeExpression(orderByExpression, sourceScope);
+                    if (orderByOnExpression) {
+                        // Try to resolve within orderByScope. If it is possible, then we warn - standard behavior differs from legacy one.
+                        for (Expression columnExpression : expressionAnalysis.getColumnReferences()) {
+                            Optional<ResolvedField> resolvedField = orderByScope.tryResolveField(columnExpression);
```

Comment:
The problem with this approach is that if you analyze `orderByExpression` with `orderByScope` in legacy mode then `ExpressionAnalysis` might throw an exception, e.g:
```
SELECT 'a' as a FROM (VALUES 1) t(a) ORDER BY a+1
```
What you probably need to do is to do column reference binding without a full expression analysis. You could achieve that by splitting `ExpressionAnalyzer` into smaller phases which could be run separately (but in proper order), e.g:
* column reference binding phase (with type binding for column references)
* type analysis for the non-reference nodes?
* ...?

This refactor would be useful since currently `ExpressionAnalyer` does both name resolution and type resolution at the same time. With such phase separation, maybe we could think about pluggable analyzers in the future (just a thought)?